### PR TITLE
[SecurityBundle] Passwords are not encoded when algorithm set to "true"

### LINKED ISSUE #34725: [Security] Passwords are not encoded when algorithm set to "true"
**Symfony version(s) affected**: 4.3.8

**Description**  

When the algorithm is set to `true`, the passwords are not encoded at all (they are stored as plaintext). I would expect to get an exception instead.

```
security:
    encoders:
        App\User\User:
            algorithm: true
```

**Additional context**  
When you mistype the `algorithm` (e.g. `bcryp` or `aut`), an exception is thrown as expected.

### PR DESCRIPTION
**Pull request by nieuwenhuisen**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #34725
| License       | MIT
| Doc PR        | -

If the algorithm is set to `true`, password will be encode as plain password.

```
security:
    encoders:
        App\User\User:
            algorithm: true
```

The reason for this is the not strict comparison of php switches.

```
switch ($config['algorithm']) {
            case 'plaintext':
}
```

`true == 'plaintext'` is `true`, so the first case is hit. My first solution was to cast the algorithm to a string, to prevent this. After some feedback I have catch this problem earlier and does not allow true as valid value to the algorithm option.

Ps. This is my first PR for Symfony, any feedback is welcome :-)!

### CODE REVIEW

**Comment by derrabus** [comment]
Can we catch this earlier? imho, `algorithm: true` shouldn't pass the config validation.

**Comment by nieuwenhuisen** [comment]
Sounds reasonable. I will take a look at the config validation.

**Comment by nieuwenhuisen** [comment]
I have reset my previous updates and change the configuration validation.
Now is true not allowed as algorithm config value.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 1]
```diff
@@ -399,7 +399,7 @@ private function addEncodersSection(ArrayNodeDefinition $rootNode)
                         ->performNoDeepMerging()
                         ->beforeNormalization()->ifString()->then(function ($v) { return ['algorithm' => $v]; })->end()
                         ->children()
-                            ->scalarNode('algorithm')->cannotBeEmpty()->end()
+                            ->scalarNode('algorithm')->cannotBeEmpty()->validate()->ifTrue()->thenInvalid('Invalid algorithm %s')->end()->end()
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: what about validating that the value is a string instead?

Comment (nieuwenhuisen) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: I can do it like this:

```
->scalarNode('algorithm')
	->cannotBeEmpty()
	->validate()
		->ifTrue(function ($v) { return !\is_string($v); })
		->thenInvalid('Invalid algorithm %s')
	->end()
->end()
```

But it is not very elegant. Is there a beter way of validating this?

Comment (nieuwenhuisen) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: Good suggestion!

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 2]
```diff
@@ -399,7 +399,13 @@ private function addEncodersSection(ArrayNodeDefinition $rootNode)
                         ->performNoDeepMerging()
                         ->beforeNormalization()->ifString()->then(function ($v) { return ['algorithm' => $v]; })->end()
                         ->children()
-                            ->scalarNode('algorithm')->cannotBeEmpty()->validate()->ifTrue()->thenInvalid('Invalid algorithm %s')->end()->end()
+                            ->scalarNode('algorithm')
+                                ->cannotBeEmpty()
+                                ->validate()
+                                    ->ifTrue(function ($v) { return false === \is_string($v); })
```
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: `->ifTrue(function ($v) { return !\is_string($v); })`

**Review by xabbuh — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Rebased on 3.4 as it applies there. Congratz for your first contrib!

**Comment by chalasr** [comment]
Thank you @nieuwenhuisen.

**Comment by mhujer** [comment]
@nieuwenhuisen Thanks for fixing it! 👍

**Pull request merged**